deepchem/hyper/random_search.py

```diff
@@ -0,0 +1,284 @@
+    or a Random Sampler from scipy.stats
+
+    Parameters
+    ----------
+    params_dict: Dict
+      Maps each hyperparameter name (string) to either a list of possible
+      parameter values or a Random Sampler from scipy.stats. If a random
+      sampler is passed, it must have the rvs method to sample values.
+    train_dataset: Dataset
+      dataset used for training
+    valid_dataset: Dataset
+      dataset used for validation (optimization on valid scores)
+    metric: Metric
+      metric used for evaluation
+    n_search: Int
+      Specifies the number of random hyperparameter searches to make.
+    output_transformers: list[Transformer]
+      Transformers for evaluation. This argument is needed since
+      `train_dataset` and `valid_dataset` may have been transformed
+      for learning and need the transform to be inverted before
+      the metric can be evaluated on a model.
+    nb_epoch: int, (default 10)
+      Specifies the number of training epochs during each iteration of optimization.
+      Not used by all model types.
+    use_max: bool, optional
+      If True, return the model with the highest score. Else return
+      model with the minimum score.
+    logdir: str, optional
+      The directory in which to store created models. If not set, will
+      use a temporary directory.
+    logfile: str, optional (default None)
+      Name of logfile to write results to. If specified, this is must
+      be a valid file name. If not specified, results of hyperparameter
+      search will be written to `logdir/results.txt`.
+
+    Returns
+    -------
+    Tuple[`best_model`, `best_hyperparams`, `all_scores`]
+      `(best_model, best_hyperparams, all_scores)` where `best_model` is
+      an instance of `dc.model.Model`, `best_hyperparams` is a
+      dictionary of parameters, and `all_scores` is a dictionary mapping
+      string representations of hyperparameter sets to validation
+      scores.
+    """
+
+    assert n_search > 0, "Number of searches should be positive"
+
+    hyperparams = params_dict.keys()
+    hyperparam_vals = params_dict.values()
+
+    # hyperparam_list should either be an Iterable sequence or a random sampler with rvs method
+    for hyperparam_list in hyperparam_vals:
+      assert isinstance(hyperparam_list, collections.abc.Iterable) \
+        or callable(getattr(hyperparam_list, "rvs", None))
+
+    if use_max:
+      best_validation_score = -np.inf
+    else:
+      best_validation_score = np.inf
+    
+    best_hyperparams = None
+    best_model = None
+    all_scores = {}
+
+    if logdir is not None:
+      if not os.path.exists(logdir):
+        os.makedirs(logdir, exist_ok=True)
+      if logfile is not None:
+        log_file = os.path.join(logdir, logfile)
+      else:
+        log_file = os.path.join(logdir, "results.txt")
+
+
+    for ind in range(n_search):
+      hyperparameter_tuple = self.get_hyp_tuple(list(hyperparam_vals))
```

Comment:
There is a chance here that the same subset of hyperparameters can explored more than once, if I am not wrong here. Maybe can we do something like get a set of all possible subset of hyperparameters via an `itertools.product` and randomly sample `n_perc` of it for searching. This may also result in dropping of `get_hyp_tuple` method.

Reply:
I think the same issue arises with continuous search ranges here